### ISSUE TITLE
How to reliably determine index of constant in OAA models

### ISSUE THREAD

**Issue by nicholastoddsmith** [VowpalWabbit/vowpal_wabbit#2676]
### Description

It would be great if the documentation could describe how constant features are mapped into the weight vector.

Using integer string features allows for easy lookup of feature weights, but the constant feature still remains challenging. I prefer not to add an "additional" column to the original data, lest it be penalized by regularization along with the rest (BTW, I assume the constant is typically not penalized in vw... am I correct?).

Using `sklearn_vw.py` I am able to recover the constant by using `CONSTANT_HASH = 116060` in binary classification problems. However, in `oaa` mode, I can't seem to figure out how the constant is mapped.

Take for instance a randomly generated multi-class problem:

```
A = np.random.rand(128, 10)
Y = (A[:, :5] * np.array([[1., 0.9, 1., 1.2, 0.8]])).argmax(1)
```
From this, it seems that the OAA models are stored at the integer indices [0...m-1] while the remaining [m...(2^b)-1] are 0, where b is the number of bits in the binary representation of m. This is apparent in the output from `invert_hash`:

```
0:0:3.90008
0[1]:1:-1.8681
0[2]:2:-0.632479
0[3]:3:-1.4639
0[4]:4:-2.31276
1:8:-1.18522
1[1]:9:3.25257
1[2]:10:-0.659727
1[3]:11:-1.2869
1[4]:12:-1.12457
2:16:-2.39304
2[1]:17:-1.82891
2[2]:18:4.50816
2[3]:19:-1.38875
2[4]:20:-0.964668
3:24:-2.59695
```

But then there is the constant. Here is with 7 bits:

```
9:72:0.0397003
9[1]:73:1.18465
9[2]:74:-1.01496
9[3]:75:-0.764425
9[4]:76:-0.621939
Constant:96:-0.233914
Constant[1]:97:-0.562953
Constant[2]:98:-0.501069
Constant[3]:99:-0.80843
Constant[4]:100:-0.772852
```
If I increase the number of bits, it remains somewhat near the end of [m...(2^b)-1] but I can't seem to find the pattern. Is there one? It would be great to write that down so people can reliably look up the constant for their models. Thanks!

### Link to Documentation Page

https://github.com/VowpalWabbit/vowpal_wabbit/wiki/Feature-Hashing-and-Extraction

**Comment by jackgerrits** [comment]
The constant feature is always added in the [constant_namespace](https://github.com/VowpalWabbit/vowpal_wabbit/blob/9918c8eb79c183bf19626ee9ebbeab0dd31b0e8e/vowpalwabbit/constant.h#L23) and with the [constant index](https://github.com/VowpalWabbit/vowpal_wabbit/blob/9918c8eb79c183bf19626ee9ebbeab0dd31b0e8e/vowpalwabbit/constant.h#L12). `CONSTANT_HASH` is the same as `constant` but with some of the topmost bits removed, which works because the default bitness is 17.

```
CONSTANT_HASH = 116060   = 000000011100010101011100
constant      = 11650396 = 101100011100010101011100
```
With OAA, there is a predictor learned for each class, which is why there is [n]... You should be able to obtain the constant weight for each predictor by taking the offset from the constant index.

I've added a [section to the wiki page](https://github.com/VowpalWabbit/vowpal_wabbit/wiki/Feature-Hashing-and-Extraction#constant-feature).

Let me know if this clears things up for you, or if it still doesn't make sense.

**Comment by nicholastoddsmith** [comment]
Thanks, that definitely clarifies things somewhat! I still can't figure out how exactly to get 96 as `(11650396 % (1 << 7)) = 92`. I'm guessing 128 from the namespace is combined with the above constant somehow and then hashed using murmur hash 3. Is that right? I tried several combinations of applying murmur hash on the namespace + constant and didn't come up with 96.

Anyway, I'm guessing it's just easier to append a constant column to problem and use that. Does VW handle the regularization any different for the constant feature it uses versus the one from this (adding a constant column) approach?

**Comment by jackgerrits** [comment]
TLDR; it will be easier to use your own constant column here. I'm quite sure VW treats the constant feature the same as others when doing regularization.

I did some digging and getting to the final value for the constant feature is a journey.
For OAA there are multiple sequential models stored internally, one per class. VW multiplies indices during example creation to make sure there is enough space for this all to be stored.

The multiplier is defined by weights per problem and stride shift. Weights per problem (`wpp`) is set to the next power of two after the number of classes. In this case number of classes is 5, so `wpp` is 8. Stride shift is used to store multiple values per weight index, in this case the value is 2. The index multiplier is:
```
multiplier = wpp << stride_shift
= 8 << 2
= 32
```

The constant feature index is multiplier by the multiplier:
```
11650396 * 32 = 372812672 
```

The index is constrained by the bitness:
```
weight mask = (2^bits << stride_shift) - 1
= (128 << 2) -1
= 511
```

This mask is then used to determine the final index:
```
final_index = (index & mask) >> stride_shift
= (372812672 & 511) >> 2
= 96
```

**Comment by nicholastoddsmith** [comment]
Thanks, this is very useful!
